### DISCUSSION TITLE
Less frequent beta builds?

### DISCUSSION THREAD

**Discussion by Atque** [Stellarium/stellarium#1513] [General]
I notice since about December, beta builds have become much more rare. Before, they were released once a week (usually each Sunday), but now it sometimes takes several weeks before a beta is released, despite some awesome new features have been introduced.

Why is that?

**Comment by alex-w** [comment]
Hi!

пт, 19 февр. 2021 г. в 16:20, Atque <notifications@github.com>:

> I notice since about December, beta builds have become much more rare.
> Before, they were released once a week (usually each Sunday), but now it
> sometimes takes several weeks before a beta is released, despite some
> awesome new features have been introduced.
>
> Why is this?
>
I have less free time since January + no frequently publishing betas now
because no need for many snapshots when code in master is almost not
changed. For example Georg introduced many changes but it was made in a
separate branch (yes, the branch has been merged now and I'll prepare new
binary packages this weekend).

-- 
With best regards, Alexander

**Reply by Atque** [reply]
Thank you for your reply. I understand, I just wondered if you might have new routines now. For me, it's no problem since I always build from source, but I guess many users would like to try some new features :)

**Reply by gzotti** [reply]
IMHO, less than you would think. Some users still report or discuss problems from 0.18... (not necessarily on GH though; at least the XP questions seem to have finally stopped :-). Most bug reports seem to come shortly after releases, not from beta testers. (Admittedly, unfortunately, sometimes the very latest features could not be tested with each and every possible way of operation in time and are added only days before release...) One problem is, shall we apply the scheme of larger projects, with "feature freeze" one month (or more) before release or so, or release new functionality as soon as it appears to provide useful results, at danger of "something again broken" or not 100% finished? Often we must work in time slots that don't respect our "seasonal" release cycle.

**Reply by Atque** [reply]
Perhaps this could be solved by making it easier for users to find the betas. Maybe some kind of obivous button saying "Check for updates"? Or the program looking for updates on startup, asking the user if (s)he wishes to update to a beta or a stable release?

**Reply by gzotti** [reply]
There is a button to check for updates in the F1/Help panel, and a "beta" button on the website. Uh, @alex-w there used to be one, why has it disappeared?  OTOH I am less happy with pushing auto-updates. For needs of occasional stargazing, 0.14 or even 0.11 will still do. Just that we cannot answer detail questions about these versions...

**Reply by Atque** [reply]
It could be more obvious, like in the first tab of the Configurations menu.

**Reply by axd1967** [reply]
here is how to reduce the impact of broken builds that reduce delays and risky new features.

currently, there is only one branch (`master`) that tracks all progress. bug fixes are applied to the `master` branch and reach users only with a new release. 

as a result, users depend on running the latest version in order to get rid of bugs; users might be waiting a long time before their bug is solved.

the usual way to deal with bugfixing in a flexible way is to create one maintenance branch per release (eg `v/20.04`). a bug is fixed in the most recent (or sometimes appropriate) maintenance branch and merged in the the `master` branch too.
every so often, when it is deemed appropriate to release the maintenance branch, a new release is created (this generates new versions/tags`v/20.04.1`, `v/20.04.2`, etc).

so, every release will have a the maintenance branch, and that branch will collect bugs as long as deemed necessary. in theory this can help users that for some reason cannot use a recent version, to receive a _limited_ amount of support for their version. it is up to the project team to decide how much releases can be supported in parallel, but that is perfectly possible with the maintenance branch approach.

this further increases the quality of Stellarium, as intermediate releases should normally only contain bugfixes, and never new - and sometimes risky - features.

this approach would dedicate the third digit of the versioning scheme (actually, the fourth... 'z' in 0.x.y.z) to bugfixing.

_ps- I'd plea again to get rid of the `0`, it is noise, there will never be a version `1` although in reality, we are long past version `1`._

**Reply by gzotti** [reply]
There is an internal target to step to 1.0 a.k.a. 2x.y, but we won't tell you here. :-) Now that Qt6 is here, this may even be another criterion to get rid of possibly outdated or useless things. But we will decide that.

**Reply by alex-w** [reply]
Alex, your proposal is not acceptable here by various reasons - at least because we use git and we follow the "rolling release" model by the fact. Your proposal was good for "classic release" model, using subversion as a CVS and tens of developers in team and this was good 15-20 years ago, but not today. I don't want spend all my free time by copying a tens identical patches between tens almost identical branches and producing a tens almost identical binary packages from these branches. 

Please do not touch versioning scheme in Stellarium, because your previous suggestions and requirements was causes ruining the building Windows packages and now we need to use two (!) versioning schemes for produce packages for all platforms.

**Reply by alex-w** [reply]
> There is a button to check for updates in the F1/Help panel, and a "beta" button on the website. Uh, @alex-w there used to be one, why has it disappeared? 

Nobody seen a "beta" button and nobody sees "weekly snapshots" link :)

**Reply by alex-w** [reply]
> It could be more obvious, like in the first tab of the Configurations menu.

Why configuration menu is more obviously place for you to checking version of program instead of Help/About menu? O_O

**Reply by Atque** [reply]
@alex-w Because that's where you save your personal configurations. I think that window is more obvious to users, than a second tab under the Help window.

**Reply by gzotti** [reply]
Matter of taste. At least one major web brower updates in the About box accessible from the Help menu.

**Comment by axd1967** [comment]
- git is perfectly fit for the multiple maintenance branch approach. see [Pro Git Book](https://git-scm.com/docs/gitworkflows#_maintenance_branch_management_after_a_feature_release) (although the explanation given there is far too complex for what I describe). a related concept from the git workflow is the [hotfix branch](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow). see also [here](https://git-scm.com/book/en/v2/Distributed-Git-Maintaining-a-Project) and [here](https://github.com/git/git/blob/master/Documentation/howto/maintain-git.txt) and [this classic](https://nvie.com/posts/a-successful-git-branching-model/).
- ten is indeed unfeasible, I did not suggest that. Keeping one or maybe two maintenance branches is perfectly feasible (it's the "_up to the project team to decide how much releases can be supported in parallel_" value). e.g. `0.20.2-maintenance` and `0.20.4-maintenance` are fine, and each time a new release is made, one older maintenance branch will be terminated; even better is to ask the community if there is a need at all to maintain more than one maintenance branch !). so there is no need to _copy_ code to different branches. one fixes a bug in the affected version, then _merges_ it where needed (into `master`, maybe in other maintenance branches).
- I'm not aware of version scheme suggestions that ruined Windows builds. (All I remember is that Windows requires a build number - in fact, Windows imposes a strict versioning scheme.).

just my 2c.
